moshi/src/main/java/com/squareup/moshi/ClassFactory.java

```diff
@@ -57,18 +59,8 @@ abstract T newInstance() throws
       // No no-args constructor. Fall back to something more magical...
     }
 
-    if (!androidChecked) {
-      try {
-        Class<?> androidBuildClass = Class.forName("android.os.Build$VERSION");
-        Field sdkIntField = androidBuildClass.getDeclaredField("SDK_INT");
-        int sdk = (int) sdkIntField.get(null);
-        if (sdk >= 29) {
-          tryUnsafe = false;
-        }
-      } catch (Exception ignored) {
-      } finally {
-        androidChecked = true;
-      }
+    if (!androidChecked && isAndroid()) {
+      tryUnsafe = isBelowAndroidApi(29);
```

Comment:
The code is setting up a complex volatile state here that is a bit hard to follow. E.g., one might wonder (as I did) why it's not just calling `isAndroid()`, since that method  *also* checks `!androidChecked` ahead of time.

What if we just sussed all these variables out at once? Like, write a method that resolves `androidChecked`, `isAndroid`, `tryUnsafe`, and `androidSdkInt` in one fell swoop. Then the code becomes easier to follow.

Reply:
Good idea - e793ecf and dcc427c